File: frontend/src/api/report.ts
/**
 * 健檢報告 API
 */
import http from './http'

// 客戶分級枚舉
export enum CustomerGrade {
  AA = 'AA',
  A = 'A',
  B = 'B',
  C = 'C'
}

// 報告生成請求
export interface ReportGenerateRequest {
  customer_id: string
  evaluation_id?: string
  format?: 'xlsx' | 'pdf'
  include_ai_analysis?: boolean
}

// 報告生成回應
export interface ReportGenerateResponse {
  report_id: string
  customer_id: string
  file_path: string
  file_format: string
  created_at: string
}

// 報告詳情
export interface ReportResponse {
  id: string
  customer_id: string
  evaluation_id?: string
  report_title: string
  report_content: Record<string, any>
  file_path?: string
  file_format: string
  created_at: string
  created_by?: string
}

// 報告列表項目
export interface ReportListItem {
  id: string
  customer_id: string
  customer_name?: string
  report_title: string
  file_format: string
  created_at: string
}

// 報告列表回應
export interface ReportListResponse {
  reports: ReportListItem[]
  total: number
  page: number
  limit: number
  total_pages: number
}

// 批次匯出請求
export interface BatchExportRequest {
  customer_ids: string[]
  format?: 'xlsx' | 'pdf'
}

// Email 發送請求
export interface ReportEmailRequest {
  report_id: string
  recipient_email: string
  subject?: string
  message?: string
}

// Email 發送回應
export interface ReportEmailResponse {
  success: boolean
  message: string
  sent_at?: string
}

/**
 * 報告 API
 */
export const reportApi = {
  /**
   * 生成健檢報告
   */
  generateReport(data: ReportGenerateRequest): Promise<ReportGenerateResponse> {
    return http.post('/api/v1/reports/generate', data)
  },

  /**
   * 獲取報告列表
   */
  getReports(params?: {
    customer_id?: string
    page?: number
    limit?: number
  }): Promise<ReportListResponse> {
    return http.get('/api/v1/reports', { params })
  },

  /**
   * 獲取單一報告詳情
   */
  getReport(reportId: string): Promise<ReportResponse> {
    return http.get(`/api/v1/reports/${reportId}`)
  },

  /**
   * 匯出報告（下載）
   */
  exportReport(reportId: string): Promise<Blob> {
    return http.get(`/api/v1/reports/${reportId}/export`, {
      responseType: 'blob'
    })
  },

  /**
   * 批次匯出報告（下載 ZIP）
   */
  batchExportReports(data: BatchExportRequest): Promise<Blob> {
    return http.post('/api/v1/reports/batch-export', data, {
      responseType: 'blob'
    })
  },

  /**
   * 透過 Email 發送報告
   */
  sendReportEmail(data: ReportEmailRequest): Promise<ReportEmailResponse> {
    return http.post('/api/v1/reports/send-email', data)
  },

  /**
   * 刪除報告
   */
  deleteReport(reportId: string): Promise<{ message: string }> {
    return http.delete(`/api/v1/reports/${reportId}`)
  },

  /**
   * 下載報告檔案（觸發瀏覽器下載）
   */
  async downloadReport(reportId: string, fileName?: string): Promise<void> {
    const blob = await this.exportReport(reportId)
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = fileName || `report_${reportId}.xlsx`
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)
  },

  /**
   * 批次下載報告（觸發瀏覽器下載 ZIP）
   */
  async downloadBatchReports(
    customerIds: string[],
    fileName?: string
  ): Promise<void> {
    const blob = await this.batchExportReports({ customer_ids: customerIds })
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = fileName || `reports_batch_${Date.now()}.zip`
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)
  }
}

File: backend/app/schemas/report.py
"""
客戶健檢報告相關的 Pydantic Schemas
"""
from pydantic import BaseModel, Field, ConfigDict, field_validator, EmailStr
from typing import Optional, List
from datetime import datetime


# ================================
# Report Generate Schemas
# ================================

class ReportGenerateRequest(BaseModel):
    """生成報告請求"""
    customer_id: str = Field(..., description="客戶 ID")
    evaluation_id: Optional[str] = Field(None, description="評估記錄 ID（不指定則使用最新）")
    format: str = Field(default="xlsx", description="報告格式")
    include_ai_analysis: bool = Field(default=True, description="包含 AI 分析結果")

    @field_validator('format')
    @classmethod
    def validate_format(cls, v: str) -> str:
        """驗證報告格式"""
        if v not in ['xlsx', 'pdf']:
            raise ValueError('報告格式只支援 xlsx 或 pdf')
        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "customer_id": "customer-123",
                "evaluation_id": None,
                "format": "xlsx",
                "include_ai_analysis": True
            }
        }
    )


class ReportGenerateResponse(BaseModel):
    """生成報告回應"""
    report_id: str = Field(..., description="報告 ID")
    customer_id: str = Field(..., description="客戶 ID")
    file_path: str = Field(..., description="檔案路徑")
    file_format: str = Field(..., description="檔案格式")
    created_at: datetime = Field(..., description="建立時間")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "report_id": "report-123",
                "customer_id": "customer-123",
                "file_path": "/storage/reports/2026-03-27/report_123.xlsx",
                "file_format": "xlsx",
                "created_at": "2026-03-27T10:00:00Z"
            }
        }
    )


# ================================
# Email Schemas
# ================================

class ReportEmailRequest(BaseModel):
    """Email 分享報告請求"""
    report_id: str = Field(..., description="報告 ID")
    recipient_email: EmailStr = Field(..., description="收件人 Email")
    subject: Optional[str] = Field(None, description="Email 主旨")
    message: Optional[str] = Field(None, description="Email 內容")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "report_id": "report-123",
                "recipient_email": "client@example.com",
                "subject": "貴公司客戶健檢報告",
                "message": "附件為貴公司的客戶健檢報告，請查收。"
            }
        }
    )


class ReportEmailResponse(BaseModel):
    """Email 發送回應"""
    success: bool = Field(..., description="是否成功")
    message: str = Field(..., description="訊息")
    sent_at: datetime = Field(..., description="發送時間")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "success": True,
                "message": "Email 發送成功",
                "sent_at": "2026-03-27T10:00:00Z"
            }
        }
    )


# ================================
# Batch Export Schemas
# ================================

class BatchExportRequest(BaseModel):
    """批次匯出報告請求"""
    customer_ids: List[str] = Field(..., description="客戶 ID 列表", min_length=1, max_length=50)
    format: str = Field(default="xlsx", description="報告格式")

    @field_validator('format')
    @classmethod
    def validate_format(cls, v: str) -> str:
        """驗證報告格式"""
        if v not in ['xlsx', 'pdf']:
            raise ValueError('報告格式只支援 xlsx 或 pdf')
        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "customer_ids": ["customer-1", "customer-2", "customer-3"],
                "format": "xlsx"
            }
        }
    )


# ================================
# Report Query Schemas
# ================================

class ReportResponse(BaseModel):
    """報告詳情回應"""
    id: str = Field(..., description="報告 ID")
    customer_id: str = Field(..., description="客戶 ID")
    evaluation_id: str = Field(..., description="評估記錄 ID")
    file_path: str = Field(..., description="檔案路徑")
    file_name: str = Field(..., description="檔案名稱")
    report_format: str = Field(..., description="報告格式")
    report_data: dict = Field(default_factory=dict, description="報告摘要資料")
    created_by: Optional[str] = Field(None, description="建立者 ID")
    created_at: datetime = Field(..., description="建立時間")
    updated_at: datetime = Field(..., description="更新時間")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": "report-123",
                "customer_id": "customer-123",
                "evaluation_id": "eval-123",
                "file_path": "/storage/reports/2026-03-27/report_123.xlsx",
                "file_name": "某公司_健檢報告_2026-03-27.xlsx",
                "report_format": "xlsx",
                "report_data": {
                    "customer_name": "某公司",
                    "grade": "A",
                    "score": 85.5,
                    "coverage_rate": 0.73
                },
                "created_by": None,
                "created_at": "2026-03-27T10:00:00Z",
                "updated_at": "2026-03-27T10:00:00Z"
            }
        }
    )


class ReportListResponse(BaseModel):
    """報告列表回應"""
    reports: List[ReportResponse] = Field(..., description="報告列表")
    total: int = Field(..., description="總數")
    page: int = Field(..., description="當前頁碼")
    limit: int = Field(..., description="每頁筆數")
    total_pages: int = Field(..., description="總頁數")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "reports": [
                    {
                        "id": "report-123",
                        "customer_id": "customer-123",
                        "evaluation_id": "eval-123",
                        "file_path": "/storage/reports/2026-03-27/report_123.xlsx",
                        "file_name": "某公司_健檢報告_2026-03-27.xlsx",
                        "report_format": "xlsx",
                        "report_data": {},
                        "created_by": None,
                        "created_at": "2026-03-27T10:00:00Z",
                        "updated_at": "2026-03-27T10:00:00Z"
                    }
                ],
                "total": 100,
                "page": 1,
                "limit": 20,
                "total_pages": 5
            }
        }
    )
